README.md
# WPExpress-UI 2.0.0

HTML Builder and Render Engine for the WPExpress Framework.

This project is no longer being maintained.

The relevant parts of this projects have been assimilated into [WPExpress Framework](https://github.com/Page-Carbajal/WPExpress).


CHANGELOG.md
# Change Log


### TO DO

- Improve Documentation
- Create Formatting class a port of WordPress functions to sanitize titles and file names
- Custom Fields
    - Create FieldsCollection/registerCustomField($fieldName, CustomField $callback) method
        - $fieldName defines the name of the field, to be created
        - CustomField is an Abstract class forcing the object to implement methods like render and so on
        - Add CustomField class
        - Samples
            - http://omarello.com/2010/08/grails-custom-tags-diy/
            - http://twig.sensiolabs.org/doc/advanced.html

### Version 2.1.0

### December 12th - Version 2.1.0 

- Created <develop> branch
- Renamed branch from master to main
- Refactored class names from .class.php to .php
- Upgraded versions of twig and mustache

            
#### BaseRenderEngine

- Implemented new RenderEngine class extending BaseRenderEngine
- Deprecated Previous RenderEngine 
- Remove Twig Support and Dependencies
- Remove Mustache Support and Dependencies
- Add BaseRenderEngine class


### Version 1.0.5 - Fixed BaseSettingsPage FieldCollection/parseFields bug

- Fixed the error of trying to parse an empty FieldCollection


### Version 1.0.4 - FieldCollection and HTMLParser

- Implemented subset for HTMLParser/parseFields
- Added sanitize_title to ID and Name properties on FieldCollection
- Added extra getters and setters for FieldCollection


### Version 1.0.3 - Input fields bug fix

- Fixed missing value on input fields


### Version 1.0.2 - Single Render Event

- Added method BaseSettingsPage/actionHookIsValid to prevent double rendering
- Simplified the constructor to allow for convention over configuration 
- Added BaseSettingsPage/setTopParentMenu

### Version 1.0.1 - Damn bugs

- Added method FieldCollection/toArray
- Renamed the method BaseSettingsPage/getFieldValue to getOptionValue


### Version 1.0.0 - Release Candidate 1

- Finished small touches on HTMLFieldParser and FieldCollection
- Fixed HTMLFieldParser/textArea method
- Added support for custom field IDs on HTMLFieldParser/parseFields
- Fixed FieldsCollection newField/type error
- Simplified BaseSettings Page
- Fixed attribution error on FieldCollection setProperty and setAttribute
- Removed field methods from BaseSettingsPage
- Added method FieldCollection/parse
- Renamed Tags to HTMLFieldParser
- Added FieldCollection class
- Renamed UI to BaseResources 
- Dropped UI folder


### Version 0.5.4 

- Added warning for empty options to Tags/selectField
- RenderEngine/createDirectoryStructure triggers a warning instead of an error 

### Version 0.5.3

- Added radio button and text area HTML tags


### Version 0.5.2 - Updated dependencies, removed tests files

- Added vendor directory and composer.lock to gitignore
- Removed the vendor directory
- Updated twig version


### Version 0.5.1 - Fixed: Exception on renderMustacheTemplate

- Added exception for cache directory on renderTwigTemplate method
- Added warning if missing partials directory
- Added warning if missing cache directory
- Fixed createDirectoryStructure exception
- Adopted [Semantic Versioning](http://semver.org)


### Version 0.5

- Fixed errors on the RenderEngine class
- Made the RenderEngine class final
- Simplified the constructor
- Added two error throwing exceptions
- Added the RenderEngine/setTemplatePath method 
- Added the RenderEngine/createDirectoryStructure method
- Remove WordPress specific functions dependencies

### Version 0.4

- Last stable version

source/FieldCollection.class.php
<?php


namespace WPExpress\UI;


class FieldCollection implements \ArrayAccess, \Countable
{

    private $container;
    private $selectedFieldName;
    private $fieldTypes = array( 'hidden', 'text', 'textarea', 'checkbox', 'radiobutton', 'select' );


    public function __construct()
    {
        $this->container         = array();
        $this->selectedFieldName = false;
    }

    private function newField( $name, $type )
    {
        $field = new \stdClass();

        $field->ID         = strtolower($name);
        $field->name       = strtolower($name);
        $field->type       = $type;
        $field->value      = '';
        $field->attributes = array();
        $field->properties = array();

        return $field;
    }

    private function addField( $name, $type = null )
    {
        if( isset( $this->container[$name] ) ) {
            // Trigger a Warning
            trigger_error("A field named {$name} is already part of the list!", E_USER_WARNING);
        }
        // Validates the given type against our list
        $fieldType               = in_array($type, $this->fieldTypes) ? $type : 'text';
        $this->container[$name]  = $this->newField($name, $fieldType);
        $this->selectedFieldName = $name;
        return $this;
    }

    public function field( $name )
    {
        $this->selectedFieldName = isset( $this->container[$name] ) ? $name : false;
        return $this;
    }

    public function addHiddenInput( $name )
    {
        $this->addField($name, 'hidden');
        return $this;
    }

    public function addTextInput( $name )
    {
        $this->addField($name, 'text');
        return $this;
    }

    public function addTextArea( $name )
    {
        $this->addField($name, 'textarea');
        return $this;
    }

    public function addRadioButton( $name )
    {
        $this->addField($name, 'radiobutton');
        return $this;
    }

    public function addCheckBox( $name )
    {
        $this->addField($name, 'checkbox');
        return $this;
    }

    public function addSelect( $name, $options )
    {
        $this->addField($name, 'select');
        $this->setProperty('options', $options);
        return $this;
    }

    public function setID( $ID )
    {
        if( $this->selectedFieldName !== false ) {
            $this->container[$this->selectedFieldName]->ID = $ID;
        }
        return $this;
    }

    public function setAttribute( $att, $value )
    {
        if( $this->selectedFieldName !== false ) {
            $this->container[$this->selectedFieldName]->attributes[$att] = $value;
        }
        return $this;
    }

    public function setAttributes( $atts )
    {
        if( $this->selectedFieldName !== false ) {
            $this->container[$this->selectedFieldName]->attributes = $atts;
        }
        return $this;
    }

    private function setProperty( $property, $value )
    {
        if( $this->selectedFieldName !== false ) {
            $this->container[$this->selectedFieldName]->properties[$property] = $value;
        }
        return $this;
    }

    //Labels and related data
    public function addLabel( $text )
    {
        $this->setProperty('label', $text);
        return $this;
    }

    public function getValue()
    {
        return isset( $this->container[$this->selectedFieldName]->properties['value'] ) ? $this->container[$this->selectedFieldName]->properties['value'] : null;
    }

    public function setValue( $value )
    {
        $this->setProperty('value', $value);
        return $this;
    }


    /****ArrayField Methods****/


    /****Parse HTML****/
    public function parseFields()
    {
        $parser = new HTMLFieldParser($this->container);
        return $parser->parseFields();
    }

    /*****Contracts*****/
    /*****Implement Interface Methods*******/
    public function offsetSet( $offset, $value )
    {
        if( !empty( $offset ) ) {
            //$this->fieldList[] = $value;
            $this->container[$offset] = $value;
        }
    }

    public function offsetExists( $offset )
    {
        return isset( $this->container[$offset] );
    }

    public function offsetUnset( $offset )
    {
        unset( $this->container[$offset] );
    }

    public function offsetGet( $offset )
    {
        return isset( $this->container[$offset] ) ? $this->container[$offset] : null;
    }

    public function count()
    {
        return count($this->container);
    }

}

source/HTMLFieldParser.class.php
<?php
/**
 * Developer: Page Carbajal (https://github.com/Page-Carbajal)
 * Date: October 21 2015, 6:30 PM
 */

namespace WPExpress\UI;


class HTMLFieldParser
{

    // This class is inspired by Groovy's FormTagLib.groovy

    private $fieldCollection;

    public function __construct( $fields )
    {
        if( !empty( $fields ) ) {
            $this->fields = $fields;
            $this->parseFields();
        }
    }

    public static function textField( $name, $attributes, $fieldProperties = null, $ID = null )
    {
        return self::renderInputField($name, 'text', $attributes, $fieldProperties, $ID);
    }

    public static function hiddenField( $name, $attributes, $fieldProperties = null, $ID = null )
    {
        return self::renderInputField($name, 'hidden', $attributes, $fieldProperties, $ID);
    }

    public static function checkboxField( $name, $attributes, $fieldProperties = null, $ID = null )
    {
        return self::renderInputField($name, 'checkbox', $attributes, $fieldProperties, $ID);
    }

    public static function radioButtonField( $name, $attributes, $fieldProperties = null, $ID = null )
    {
        return self::renderInputField($name, 'radio', $attributes, $fieldProperties, $ID);
    }

    private static function renderInputField( $name, $type, $attributes, $fieldProperties, $ID )
    {
        $controlID       = empty( $ID ) ? $name : $ID;
        $baseAttributes  = array( 'type' => $type, 'id' => $controlID, 'name' => $name );
        $attString       = self::arrayToHTMLAttributes(array_merge($baseAttributes, $attributes));
        $fieldProperties = empty( $fieldProperties ) ? array() : ( is_array($fieldProperties) ? $fieldProperties : array( $fieldProperties ) );

        $field             = new \stdClass();
        $field->html       = "<input {$attString} />";
        $field->properties = array_merge($baseAttributes, $attributes, $fieldProperties);

        return $field;
    }

    protected static function arrayToHTMLAttributes( $list, $includeValue = true )
    {
        $attributes = array_map(function ( $key ) use ( $list, $includeValue ) {
            if( is_bool($list[$key]) ) {
                return ( $list[$key] ? $key : '' );
            }
            if( $key == 'value' && ( !$includeValue ) ) {
                return ""; //if value should not be rendered as an attribute
            }
            $propertyValue = esc_html($list[$key]);
            return "{$key}=\"{$propertyValue}\"";
        }, array_keys($list));
        return implode(' ', $attributes);
    }

    /**
     * Generate a Select Field.
     * @param $name
     * @param $options : An array with items. Accepts simple array or nested array with the properties [text, value, selected]
     * @param $attributes
     * @return bool|string
     */
    public static function selectField( $name, $options, $attributes, $value = null, $fieldProperties = null, $ID = null )
    {
        if( empty( $name ) ) {
            return false;
        }

        if( empty( $options ) ) {
            trigger_error("Object <strong>{$name}</strong> has no valid options!", E_USER_WARNING);
            return false;
        }

        $fieldProperties   = empty( $fieldProperties ) ? array() : ( is_array($fieldProperties) ? $fieldProperties : array( $fieldProperties ) );
        $baseAttributes    = array( 'id' => $name, 'name' => $name );
        $source            = '';
        $hasSelectedOption = false;

        foreach( $options as $item ) {
            $option         = '';
            $selectedOption = '';

            if( !$hasSelectedOption && ( isset( $item['selected'] ) || ( $item == $value ) || ( isset( $item['value'] ) && $value == $item['value'] ) ) ) {
                $selectedOption    = 'selected="selected"';
                $hasSelectedOption = true;
            }

            if( !is_array($item) ) {
                $option .= "<option {$selectedOption} value=\"{$item}\">{$item}</option>";
            } else {
                $option = "<option ";
                if( isset( $item['value'] ) ) {
                    $option .= " value=\"{$item['value']}\" {$selectedOption}";
                } else {
                    $option .= " value=\"{$item['text']}\" {$selectedOption}";
                }

                $option = trim($option) . ">{$item['text']}</option>";
            }
            $source .= $option;
        }

        $attsString = self::arrayToHTMLAttributes(array_merge($baseAttributes, $attributes));
        $output     = "<select {$attsString}>{$source}</select>";

        $field             = new \stdClass();
        $field->html       = $output;
        $field->properties = array_merge($baseAttributes, $attributes, array( 'options' => $options ), $fieldProperties);

        return $field;
    }

    public static function textArea( $name, $value, $attributes, $fieldProperties = null, $ID = null )
    {
        $controlID       = empty( $ID ) ? $name : $ID;
        $fieldProperties = empty( $fieldProperties ) ? array() : ( is_array($fieldProperties) ? $fieldProperties : array( $fieldProperties ) );
        $baseAttributes  = array( 'name' => $name, 'id' => $controlID );
        $attsString      = self::arrayToHTMLAttributes(array_merge($baseAttributes, $attributes));

        $field             = new \stdClass();
        $field->html       = "<textarea {$attsString}>{$value}</textarea>";
        $field->properties = array_merge($baseAttributes, $attributes, $fieldProperties);

        return $field;
    }


    public function parseFields()
    {
        $list = array();

        foreach( $this->fields as $field ) {

            switch( $field->type ) {
                case "select":
                    // Why not simply pass the field object? Well, we could, but we'll be forcing you to use our FieldCollection. So we keep it cool.
                    $list[] = $this->selectField($field->name, $field->properties['options'], $field->attributes, $field->properties['value'], $field->properties, $field->ID);
                    break;
                case "radio":
                    $list[] = $this->radioButtonField($field->name, $field->attributes, $field->properties, $field->ID);
                    break;
                case "radiobutton":
                    $list[] = $this->radioButtonField($field->name, $field->attributes, $field->properties, $field->ID);
                    break;
                case "check":
                    $list[] = $this->checkboxField($field->name, $field->attributes, $field->properties, $field->ID);
                    break;
                case "checkbox":
                    $list[] = $this->checkboxField($field->name, $field->attributes, $field->properties, $field->ID);
                    break;
                case "textarea":
                    $list[] = $this->textArea($field->name, $field->properties['value'], $field->attributes, $field->properties, $field->ID);
                    break;
                default:
                    $list[] = $this->textField($field->name, $field->attributes, $field->properties, $field->ID);
                    break;
            }

        }

        return $list;
    }

}

source/UI/RenderEngine.class.php
<?php


namespace WPExpress\UI;


use Mustache_Engine;
use Mustache_Loader_FilesystemLoader;
use Twig_Loader_Filesystem;
use Twig_Environment;


final class RenderEngine
{
    protected $type;
    protected $typeExtension;
    protected $templatePath;
    protected $templateFolder;
    protected $useTypeAsExtension;

    public function __construct( $templateFolderPath, $type = 'mustache', $useTypeAsExtension = true )
    {
        $this->type               = trim(strtolower($type));
        $this->useTypeAsExtension = $useTypeAsExtension;
        $this->typeExtension      = sanitize_title($this->type); // TODO: Port this function to Plain PHP
        $this->setTemplatePath($templateFolderPath)->createDirectoryStructure();
    }

    public function setTemplatePath( $templateFolderPath )
    {
        if( ( !empty( $templateFolderPath ) ) && file_exists($templateFolderPath) ) {
            $this->templateFolder = $templateFolderPath;
        } else {
            throw new \Exception('Error: Template path does not exist! - WPExpress/UI @ RenderEngine');
        }

        return $this;
    }

    public function setTypeAsExtension( $flag )
    {
        $this->useTypeAsExtension = $flag;
    }

    private function createDirectoryStructure()
    {
        $success = true;

        // Create the directory <path-to-template>/mustache/partials
        $cacheDirectory = $this->getBaseDirectory() . "/cache/";
        if( !file_exists($cacheDirectory) ) {
            $success = $this->makeDirectoryFromPath($cacheDirectory);
        }

        $cachePath = $this->getBaseDirectory() . "/partials/";
        if( !file_exists($cachePath) ) {
            $success = $this->makeDirectoryFromPath($cachePath);
        }

        if( !$success ) {
            trigger_error('Error: Cant write in the specified template path. Check for permissions. - WPExpress/UI @ RenderEngine', E_USER_WARNING);
        }

        return $this;
    }

    // This function is a verbatim copy of WordPress wp_is_stream
    private function isURLStream( $path )
    {
        $wrappers    = stream_get_wrappers();
        $wrappers_re = '(' . join('|', $wrappers) . ')';

        return preg_match("!^$wrappers_re://!", $path) === 1;
    }

    // This function is a verbatim copy of WordPress wp_mkdir_p
    private function makeDirectoryFromPath( $target )
    {

        $wrapper = null;

        // Strip the protocol.
        if( $this->isURLStream($target) ) {
            list( $wrapper, $target ) = explode('://', $target, 2);
        }

        // From php.net/mkdir user contributed notes.
        $target = str_replace('//', '/', $target);

        // Put the wrapper back on the target.
        if( $wrapper !== null ) {
            $target = $wrapper . '://' . $target;
        }

        /*
         * Safe mode fails with a trailing slash under certain PHP versions.
         * Use rtrim() instead of untrailingslashit to avoid formatting.php dependency.
         */
        $target = rtrim($target, '/');
        if( empty( $target ) )
            $target = '/';

        if( file_exists($target) )
            return @is_dir($target);

        // We need to find the permissions of the parent folder that exists and inherit that.
        $target_parent = dirname($target);
        while( '.' != $target_parent && !is_dir($target_parent) ) {
            $target_parent = dirname($target_parent);
        }

        // Get the permission bits.
        if( $stat = @stat($target_parent) ) {
            $dir_perms = $stat['mode'] & 0007777;
        } else {
            $dir_perms = 0777;
        }

        if( @mkdir($target, $dir_perms, true) ) {

            /*
             * If a umask is set that modifies $dir_perms, we'll have to re-set
             * the $dir_perms correctly with chmod()
             */
            if( $dir_perms != ( $dir_perms & ~umask() ) ) {
                $folder_parts = explode('/', substr($target, strlen($target_parent) + 1));
                for( $i = 1; $i <= count($folder_parts); $i++ ) {
                    @chmod($target_parent . '/' . implode('/', array_slice($folder_parts, 0, $i)), $dir_perms);
                }
            }

            return true;
        }

        return false;

    }

    public function getBaseDirectory()
    {
        return untrailingslashit($this->templateFolder) . "/{$this->type}/";
    }

    private function parseFileName( $fileName )
    {
        return ( $this->useTypeAsExtension ? "{$fileName}.{$this->typeExtension}" : $fileName );
    }

    public function getTemplatePath( $fileName )
    {
        $pathToFile = trailingslashit($this->getBaseDirectory()) . $this->parseFileName($fileName);
        if( file_exists($pathToFile) ) {
            return $pathToFile;
        }

        return false;
    }

    public function renderTemplate( $fileName, $context )
    {
        $template = $this->getTemplatePath($fileName);
        if( $template !== false ) {
            switch( $this->type ) {
                case "twig":
                    $raw = $this->renderTwigTemplate($fileName, $context);
                    break;
                default:
                    $raw = $this->renderMustacheTemplate($fileName, $context);
                    break;
            }
            return $raw;
        }

        // TODO: Improve this message
        return "<strong>Not it!</strong>";
    }

    private function renderMustacheTemplate( $fileName, $context )
    {
        $options = array();
        $options['loader'] = new Mustache_Loader_FilesystemLoader($this->getBaseDirectory());
        $options['charset'] = 'UTF-8';

        if( file_exists($this->getBaseDirectory() . '/cache') ) {
            $options['cache'] = $this->getBaseDirectory() . '/cache';
        } else {
            trigger_error('No cache directory has been defined for Mustache templates. WPExpress/UI @ RenderEngine');
        }

        if( file_exists($this->getBaseDirectory() . '/partials') ) {
            $options['partials_loader'] = new Mustache_Loader_FilesystemLoader($this->getBaseDirectory() . '/partials');
        } else {
            trigger_error('No partials directory has been defined for Mustache templates. WPExpress/UI @ RenderEngine');
        }

        if( function_exists('apply_filters') ) {
            // Prevent the use of WordPress specific functions
            $options = apply_filters('wpex_set_mustache_engine_options', $options);
        }

        $mustache = new Mustache_Engine($options);


        return $mustache->render($fileName, $context);
    }

    private function renderTwigTemplate( $fileName, $context )
    {
        if( !file_exists( $this->getBaseDirectory() . '/cache' ) ){
            throw new \Exception( 'No cache directory has been defined for Twig. - WPExpress/UI @ Render Engine.', 404 );
        }

        $loader   = new Twig_Loader_Filesystem($this->getBaseDirectory());
        $twig     = new Twig_Environment($loader, array( 'cache' => $this->getBaseDirectory() . '/cache' ));
        $fileName = $this->parseFileName($fileName);
        $template = $twig->loadTemplate($fileName);
        return $template->render($context);
    }

}

source/UI/HTML/Tags.class.php
<?php
/**
 * Developer: Page Carbajal (https://github.com/Page-Carbajal)
 * Date: October 21 2015, 6:30 PM
 */

namespace WPExpress\UI\HTML;



class Tags
{

    // This class is inspired by Groovy's FormTagLib.groovy
    // https://github.com/grails/grails-core/blob/0b1d6a6d02f2217643a69e8314f76078dacbce32/grails-plugin-gsp/src/main/groovy/org/grails/plugins/web/taglib/FormTagLib.groovy

    public function __construct()
    {

    }

    public static function textField($name, $attributes)
    {
        $defaultProperties = array(
            'type' => 'text',
            'id' => $name,
            'name' => $name
        );

        return self::renderInputField(array_merge( $defaultProperties, $attributes ));
    }

    public static function hiddenField($name, $attributes)
    {
        $defaultProperties = array(
            'type' => 'hidden',
            'id' => $name,
            'name' => $name,
        );

        return self::renderInputField(array_merge( $defaultProperties, $attributes ));
    }

    public static function checkboxField($name, $attributes)
    {
        $defaultProperties = array(
            'type' => 'checkbox',
            'id' => $name,
            'name' => $name,
        );

        return self::renderInputField(array_merge( $defaultProperties, $attributes ));
    }

    public static function radioButtonField($name, $attributes)
    {
        $defaultProperties = array(
            'type' => 'radio',
            'name' => $name,
        );

        return self::renderInputField(array_merge( $defaultProperties, $attributes ));
    }

    protected static function renderInputField($attributes)
    {
        $atts = self::arrayToHTMLAttributes( $attributes );

        $field = new \stdClass();
        $field->html = "<input {$atts} />";
        $field->properties = $attributes;

        return $field;
    }

    protected static function arrayToHTMLAttributes($list, $includeValue = true)
    {
        $attributes = array_map(function($key) use ($list, $includeValue)
        {
            if( is_bool($list[$key]) )
            {
                return ( $list[$key] ? $key : '');
            }
            if($key == 'value' && (!$includeValue)){
                return ""; //if value should not be rendered as an attribute
            }
            $propertyValue = esc_html( $list[$key] );
            return "{$key}=\"{$propertyValue}\"";
        },  array_keys($list) );
        return implode( ' ', $attributes );
    }

    /**
     * Generate a Select Field.
     * @param $name
     * @param $options: An array with items. Accepts simple array or nested array with the properties [text, value, selected]
     * @param $attributes
     * @return bool|string
     */
    public static function selectField($name, $options, $attributes)
    {
        if( empty($name) ){
            return false;
        }

        if( empty($options) ){
            trigger_error("Object <strong>{$name}</strong> has no valid options!", E_USER_WARNING);
            return false;
        }

        $properties = array( 'id' => $name, 'name' => $name );
        $source = '';

        foreach( $options as $item ){
            $option = '';
            if( !is_array($item) ){
                $option .= "<option value=\"{$item}\">{$item}</option>";
            } else {
                $option = "<option ";
                if( isset($item['value']) ){
                    $option .= " value=\"{$item['value']}\"";
                } else {
                    $option .= " value=\"{$item['text']}\"";
                }

                if( isset($item['selected']) && $item['selected'] == true ){
                    $option .= ' selected="selected" ';
                }
                $option = trim( $option ) . ">{$item['text']}</option>";
            }
            $source .= $option;
        }

        $attsString = self::arrayToHTMLAttributes( array_merge($properties, $attributes) );
        $output = "<select {$attsString}>{$source}</select>";

        $field = new \stdClass();
        $field->html = $output;
        $field->properties = array_merge( $properties, $attributes, array('options' => $options) );

        return $field;
    }

    public static function textArea( $name, $value, $attributes )
    {
        $properties = array( 'id' => $name, 'name' => $name );
        $attsString = self::arrayToHTMLAttributes( array_merge($properties, $attributes) );
        $output = "<textarea {$attsString}>{$value}</textarea>";

        return $output;
    }


    public static function parseFields($collection)
    {
        $list = array();

        foreach( $collection as $item ){

            switch($item['type']){
                case "select":
                    $list[] = self::selectField($item['name'], $item['options'], $item['properties']);
                    break;
                case "radio":
                    $list[] = self::radioButtonField( $item['name'], $item['properties'] );
                    break;
                case "radiobutton":
                    $list[] = self::radioButtonField( $item['name'], $item['properties'] );
                    break;
                case "check":
                    $list[] = self::checkboxField($item['name'], $item['properties']);
                    break;
                case "checkbox":
                    $list[] = self::checkboxField($item['name'], $item['properties']);
                    break;
                case "textarea":
                    $list[] = self::textArea($item['name'], $item['value'], $item['properties']);
                    break;
                default:
                    $list[] = self::textField($item['name'], $item['properties']);
                    break;
            }

        }

        return $list;
    }

}